src/app/digital/internal/DigitalComponents.ts

```diff
@@ -285,9 +285,9 @@ class DigitalComponentInfoProvider implements ObjInfoProvider {
     private readonly ports: Map<string, ObjInfo>;
 
     public constructor(components: DigitalComponentInfo[]) {
-        this.components = new Map(components.map((info) => [info.kind, info]));
-        this.wires = new Map([["DigitalWire", WireInfo]]);
-        this.ports = new Map([["DigitalPort", PortInfo]]);
+        this.components = new Map<string, ComponentInfo>(components.map((info) => [info.kind, info]));
```

Comment:
These explicit types were required.  I'm not sure how the type inference was working before, because the error messages I was getting implied RHS is a `Map<string, DigitalComponentInfo>`, which was being assigned to a `Map<string, ComponentInfo>`.  This shouldn't be a compatible assignment...

The error messages were complaining about the new `upsert` method, like:
```
error TS2322: Type 'Map<string, DigitalComponentInfo>' is not assignable to type 'Map<string, ComponentInfo>'.
      Types of property 'upsert' are incompatible.
        Types of parameters 'updateFn' and 'updateFn' are incompatible.
          Type 'ComponentInfo' is not assignable to type 'DigitalComponentInfo'.
```
Do you have any insights into how to coerce the `upsert` definition so the RHS infers its type from the LHS instead of from its constructor args?